### PR TITLE
refactor: Replace `http` with `fetch` - 2

### PR DESCRIPTION
**Pull request by KevLehman**

### CODE REVIEW

File: apps/meteor/app/custom-oauth/server/custom_oauth_server.js [hunk 1]
```diff
@@ -130,7 +130,9 @@ export class CustomOAuth {
 		}
 
 		try {
-			response = HTTP.post(this.tokenPath, allOptions);
+			const queryparams = new URLSearchParams(allOptions.params);
+			const request = await fetch(`${this.tokenPath}?${queryparams.toString()}`, { method: 'POST', headers: allOptions.headers });
```
Comment (rodrigok) [inline apps/meteor/app/custom-oauth/server/custom_oauth_server.js:134]: You are missing the `allOptions.auth` that is set above.

File: apps/meteor/app/custom-oauth/server/custom_oauth_server.js [hunk 2]
```diff
@@ -130,7 +130,9 @@ export class CustomOAuth {
...
+			response = await request.json();
```
Comment (rodrigok) [inline apps/meteor/app/custom-oauth/server/custom_oauth_server.js:135]: Response from `HTTP.post` is different from the `json()` result right? I believe it's the data only and the response variable used below is trying to get data from `response.data`.

File: apps/meteor/app/custom-oauth/server/custom_oauth_server.js [hunk 3]
```diff
@@ -165,10 +167,9 @@ export class CustomOAuth {
 		}
 
 		try {
-			const response = HTTP.get(this.identityPath, {
-				headers,
-				params,
-			});
+			const queryparams = new URLSearchParams(params);
+			const request = await fetch(`${this.identityPath}?${queryparams.toString()}`, { method: 'GET', headers });
+			const response = await request.json();
```
Comment (rodrigok) [inline apps/meteor/app/custom-oauth/server/custom_oauth_server.js:172]: Same issue with the response

File: apps/meteor/app/livechat/server/api/v1/webhooks.ts [hunk 4]
```diff
@@ -58,7 +57,7 @@ API.v1.addRoute(
 				headers: {
 					'X-RocketChat-Livechat-Token': settings.get<string>('Livechat_secret_token'),
 				},
-				data: sampleData,
+				body: JSON.stringify(sampleData),
```
Comment (rodrigok) [inline apps/meteor/app/livechat/server/api/v1/webhooks.ts:60]: I believe the post with json needs to pass the content type header as well

**Review by rodrigok — CHANGES_REQUESTED** [review]

File: packages/server-fetch/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	},
+	allowSelfSignedCerts?: boolean,
+): ReturnType<typeof fetch> {
+	const agent = getFetchAgent(input, allowSelfSignedCerts);
+
+	const controller = new AbortController();
+	const timeoutId = setTimeout(() => controller.abort(), options?.timeout ?? 20000);
+
+	const params = new URLSearchParams(options?.params).toString();
+	const url = `${input}${params ? `?${params}` : ''}`;
```
Comment (rodrigok) [inline packages/server-fetch/src/index.ts:64]: Here the `input` may already have the `?` so my suggestion is:
1. use URL to parse the input as `url = new URL(input)`
2. loop the params and use `url.searchParams.add(key, value)`
3. use `url.toString()` or `url.href` to get the final URL.

Then we allow merge of params

File: packages/server-fetch/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+	if (isPostOrPutWithBody(options)) {
+		try {
+			// Try to parse body as JSON
+			JSON.parse(options?.body as string);
```
Comment (rodrigok) [inline packages/server-fetch/src/index.ts:69]: Should not parse the body, but test if it's an object or array, if so then stringify and add the header (being able to override as you did). If someone pass as string they should inform the header

**Review by rodrigok — CHANGES_REQUESTED** [review]

File: apps/meteor/server/services/omnichannel-integrations/providers/voxtelesys.ts [hunk 7]
```diff
@@ -133,14 +133,15 @@ export class Voxtelesys implements ISMSProvider {
 
 		const options = {
 			headers: {
-				Authorization: `Bearer ${this.authToken}`,
+				'Authorization': `Bearer ${this.authToken}`,
+				'Content-Type': 'application/json',
```
Comment (rodrigok) [inline apps/meteor/server/services/omnichannel-integrations/providers/voxtelesys.ts:137]: Not necessary

File: apps/meteor/server/services/nps/sendNpsResults.ts [hunk 8]
```diff
@@ -23,9 +23,10 @@ export const sendNpsResults = async function sendNpsResults(npsId: string, data:
 			await fetch(`${npsUrl}/v1/surveys/${npsId}/results`, {
 				method: 'POST',
 				headers: {
-					Authorization: `Bearer ${token}`,
+					'Authorization': `Bearer ${token}`,
+					'Content-Type': 'application/json',
```
Comment (rodrigok) [inline apps/meteor/server/services/nps/sendNpsResults.ts:27]: Not necessary

File: apps/meteor/server/cron/statistics.js [hunk 9]
```diff
@@ -15,7 +15,7 @@ async function generateStatistics(logger) {
 	}
 
 	try {
-		const headers = {};
+		const headers = { 'Content-Type': 'application/json' };
```
Comment (rodrigok) [inline apps/meteor/server/cron/statistics.js:18]: Not necessary

File: apps/meteor/ee/server/apps/marketplace/appInstall.ts [hunk 10]
```diff
@@ -1,14 +1,14 @@
...
 import { getWorkspaceAccessToken } from '../../../../app/cloud/server';
 import { settings } from '../../../../app/settings/server';
 import { Info } from '../../../../app/utils/server';
-import { fetch } from '../../../../server/lib/http/fetch';
 
 type installAction = 'install' | 'update' | 'uninstall';
 
 export async function notifyAppInstall(marketplaceBaseUrl: string, action: installAction, appInfo: IAppInfo): Promise<void> {
-	const headers: { Authorization?: string } = {};
+	const headers: { 'Authorization'?: string; 'Content-Type': string } = { 'Content-Type': 'application/json' };
```
Comment (rodrigok) [inline apps/meteor/ee/server/apps/marketplace/appInstall.ts:11]: Not necessary

File: apps/meteor/ee/server/apps/appRequestsCron.ts [hunk 11]
```diff
@@ -29,13 +29,14 @@ const appsNotifyAppRequests = async function _appsNotifyAppRequests() {
 
 		const options = {
 			headers: {
-				Authorization: `Bearer ${token}`,
+				'Authorization': `Bearer ${token}`,
+				'Content-Type': 'application/json',
```
Comment (rodrigok) [inline apps/meteor/ee/server/apps/appRequestsCron.ts:33]: Not necessary

**Review by rodrigok — CHANGES_REQUESTED** [review]

File: apps/meteor/app/federation/server/lib/http.js [hunk 12]
```diff
@@ -19,11 +19,17 @@ export async function federationRequest(method, url, body, headers, peerKey = nu
...
-		headers: { ...headers, 'x-federation-domain': getFederationDomain() },
-	});
+	try {
+		const request = await fetch(url, {
+			method,
+			headers: { ...headers, 'x-federation-domain': getFederationDomain() },
+			body: data,
+			timeout: 2000,
+		});
+		return request.json();
```
Comment (rodrigok) [inline apps/meteor/app/federation/server/lib/http.js:29]: What about the places calling this method since the result now is different?

**Review by rodrigok — DISMISSED** [review]

File: apps/meteor/app/custom-oauth/server/custom_oauth_server.js [hunk 13]
```diff
@@ -123,35 +123,32 @@ export class CustomOAuth {
 
 		// Only send clientID / secret once on header or payload.
 		if (this.tokenSentVia === 'header') {
-			allOptions.auth = `${config.clientId}:${OAuth.openSecret(config.secret)}`;
+			allOptions.headers.Authorization = Buffer.from(`${config.clientId}:${OAuth.openSecret(config.secret)}`).toString('base64');
```
Comment (MarcosSpessatto) [inline apps/meteor/app/custom-oauth/server/custom_oauth_server.js:126]: Aren't we forgetting to concatenate a `Basic ${base64}` here together with the base64 string, informing which type of auth we are sending?

File: apps/meteor/app/custom-oauth/server/custom_oauth_server.js [hunk 14]
```diff
@@ -123,35 +123,32 @@ export class CustomOAuth {
...
 		} catch (err) {
 			const error = new Error(`Failed to complete OAuth handshake with ${this.name} at ${this.tokenPath}. ${err.message}`);
 			throw _.extend(error, { response: err.response });
 		}
 
-		let data;
-		if (response.data) {
-			data = response.data;
-		} else {
-			data = JSON.parse(response.content);
```
Comment (MarcosSpessatto) [inline apps/meteor/app/custom-oauth/server/custom_oauth_server.js:143]: Looks like it was also expecting some sort of text response here, isn't? 🤔

File: apps/meteor/app/custom-oauth/server/custom_oauth_server.js [hunk 15]
```diff
@@ -165,22 +162,12 @@ export class CustomOAuth {
...
-				headers,
-				params,
-			});
-
-			let data;
-
-			if (response.data) {
-				data = response.data;
-			} else {
-				data = JSON.parse(response.content);
```
Comment (MarcosSpessatto) [inline apps/meteor/app/custom-oauth/server/custom_oauth_server.js:178]: Same question as above 🤔

File: apps/meteor/app/federation/server/lib/http.js [hunk 16]
```diff
@@ -19,11 +19,17 @@ export async function federationRequest(method, url, body, headers, peerKey = nu
 
 	httpLogger.debug(`[${method}] ${url}`);
 
-	return MeteorHTTP.call(method, url, {
-		data,
-		timeout: 2000,
-		headers: { ...headers, 'x-federation-domain': getFederationDomain() },
-	});
+	try {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/federation/server/lib/http.js:22]: Is this try/catch necessary since we are just rethrowing the error?

File: apps/meteor/app/livechat/server/api/v1/webhooks.ts [hunk 17]
```diff
@@ -69,10 +71,11 @@ API.v1.addRoute(
 
 			try {
 				Livechat.logger.debug(`Testing webhook ${webhookUrl}`);
-				const response = HTTP.post(webhookUrl, options);
+				const request = await fetch(webhookUrl, options);
+				const response = await request.text();
```
Comment (MarcosSpessatto) [inline apps/meteor/app/livechat/server/api/v1/webhooks.ts:75]: If we are expecting a text response, shouldn't we use text within the `Accept` header instead of `application/json`?

File: apps/meteor/app/livechat/server/methods/webhookTest.ts [hunk 18]
```diff
@@ -73,15 +73,17 @@ Meteor.methods<ServerMethods>({
...
+				'Accept': 'application/json',
 			},
-			body: JSON.stringify(sampleData),
+			body: sampleData,
 		};
 
 		const response = await postCatchError(settings.get('Livechat_webhookUrl'), options);
 
-		SystemLogger.debug({ response });
+		SystemLogger.debug({ response: await response?.text() });
```
Comment (MarcosSpessatto) [inline apps/meteor/app/livechat/server/methods/webhookTest.ts:86]: Same as the previous question/comment.

File: apps/meteor/app/slackbridge/server/SlackAPI.js [hunk 19]
```diff
@@ -1,14 +1,13 @@
...
 	constructor(apiToken) {
 		this.apiToken = apiToken;
 	}
 
-	getChannels(cursor = null) {
+	async getChannels(cursor = null) {
 		let channels = [];
-
-		const response = HTTP.get('https://slack.com/api/conversations.list', {
+		const request = await fetch(`https://slack.com/api/conversations.list`, {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/slackbridge/server/SlackAPI.js:10]: Don't need to be template strings

File: apps/meteor/app/slackbridge/server/SlackAPI.js [hunk 20]
```diff
@@ -17,21 +16,22 @@ export class SlackAPI {
...
 		}
 
 		return channels;
 	}
 
-	getGroups(cursor = null) {
+	async getGroups(cursor = null) {
 		let groups = [];
-		const response = HTTP.get('https://slack.com/api/conversations.list', {
+		const request = await fetch(`https://slack.com/api/conversations.list`, {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/slackbridge/server/SlackAPI.js:34]: Same

File: apps/meteor/app/slackbridge/server/SlackAPI.js [hunk 21]
```diff
@@ -40,102 +40,133 @@ export class SlackAPI {
...
 			}
 		}
 
 		return groups;
 	}
 
-	getRoomInfo(roomId) {
-		const response = HTTP.get('https://slack.com/api/conversations.info', {
+	async getRoomInfo(roomId) {
+		const request = await fetch(`https://slack.com/api/conversations.info`, {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/slackbridge/server/SlackAPI.js:57]: Same

File: apps/meteor/app/slackbridge/server/SlackAPI.js [hunk 22]
```diff
@@ -40,102 +40,133 @@ export class SlackAPI {
...
-	getMembers(channelId) {
+	async getMembers(channelId) {
 		const { num_members } = this.getRoomInfo(channelId);
 		const MAX_MEMBERS_PER_CALL = 100;
 		let members = [];
 		let currentCursor = '';
 		for (let index = 0; index < num_members; index += MAX_MEMBERS_PER_CALL) {
-			const response = HTTP.get('https://slack.com/api/conversations.members', {
+			// eslint-disable-next-line no-await-in-loop
+			const request = await fetch(`https://slack.com/api/conversations.members`, {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/slackbridge/server/SlackAPI.js:75]: Same

File: apps/meteor/app/slackbridge/server/SlackAPI.js [hunk 23]
```diff
@@ -40,102 +40,133 @@ export class SlackAPI {
...
 		});
-		return response && response.data;
+		const response = await request.json();
+		return response;
 	}
 
-	getPins(channelId) {
-		const response = HTTP.get('https://slack.com/api/pins.list', {
+	async getPins(channelId) {
+		const request = await fetch(`https://slack.com/api/pins.list`, {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/slackbridge/server/SlackAPI.js:152]: Same

File: apps/meteor/app/slackbridge/server/SlackAPI.js [hunk 24]
```diff
@@ -40,102 +40,133 @@ export class SlackAPI {
...
 		});
-		return response && response.data && response.statusCode === 200 && response.data.ok && response.data.items;
+		const response = await request.json();
+		return response && response && request.status === 200 && request.ok && response.items;
 	}
 
-	getUser(userId) {
-		const response = HTTP.get('https://slack.com/api/users.info', {
+	async getUser(userId) {
+		const request = await fetch(`https://slack.com/api/users.info`, {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/slackbridge/server/SlackAPI.js:163]: Same

File: apps/meteor/app/slackbridge/server/SlackAPI.js [hunk 25]
```diff
@@ -40,102 +40,133 @@ export class SlackAPI {
...
-			params: Object.assign({ token: this.apiToken }, options),
+	async getHistory(family, options) {
+		const request = await fetch(`https://slack.com/api/${family}.history`, {
+			params: {
+				token: this.apiToken,
+				...options,
+			},
 		});
-		return response && response.data;
+		const response = await request.json();
```
Comment (MarcosSpessatto) [inline apps/meteor/app/slackbridge/server/SlackAPI.js:147]:
```suggestion
		return request.json();
```

File: apps/meteor/ee/server/apps/communication/rest.ts [hunk 26]
```diff
@@ -1219,9 +1225,18 @@ export class AppsRestApi {
 					const { unseenRequests } = this.bodyParams;
 
 					try {
-						const result = HTTP.post(`${baseUrl}/v1/app-request/markAsSeen`, { headers, data: { ids: unseenRequests } });
+						const request = await fetch(`${baseUrl}/v1/app-request/markAsSeen`, {
+							method: 'POST',
+							headers,
+							body: JSON.stringify({ ids: unseenRequests }),
```
Comment (MarcosSpessatto) [inline apps/meteor/ee/server/apps/communication/rest.ts:1231]: Not sure if this (stringify) is necessary anymore

File: apps/meteor/server/services/omnichannel-integrations/providers/mobex.ts [hunk 27]
```diff
@@ -119,11 +119,17 @@ export class Mobex implements ISMSProvider {
...
+					username: currentUsername,
+					password: currentPassword,
+					to: strippedTo,
+					from: currentFrom,
+					content: message,
+				},
+			});
 
-			const json = await response.json();
+			const json = await response.text();
```
Comment (MarcosSpessatto) [inline apps/meteor/server/services/omnichannel-integrations/providers/mobex.ts:132]: Why did you change this to text? 🤔

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]
@rodrigok I saw your comments asking to remove the `application/json` headers from all the requests, but according to the parser, it needs to be there, otherwise, it won't be parsed as a JSON request.

https://github.com/RocketChat/Rocket.Chat/pull/28828/files#diff-db5d599446e75326c7cbd66121cdfdd3e97d0d35ff225f2aa4d4216f1a82f854R41

Or am I missing something? 🤔

File: apps/meteor/app/autotranslate/server/googleTranslate.ts [hunk 28]
```diff
@@ -85,37 +85,35 @@ class GoogleAutoTranslate extends AutoTranslate {
...
+		let result: { data?: { languages: ISupportedLanguage[] } } = {};
 		const params = {
 			key: this.apiKey,
 			...(target && { target }),
 		};
 
 		try {
-			const request = await fetch(`https://translation.googleapis.com/language/translate/v2/languages?${new URLSearchParams(params)}`);
-			result = await request.json();
+			const request = await fetch(`https://translation.googleapis.com/language/translate/v2/languages`, { params });
```
[the remark this answers is not in the document]
Comment (KevLehman) [reply apps/meteor/app/autotranslate/server/googleTranslate.ts:95]: Well, this keeps the "original" logic so won't change it

File: apps/meteor/app/autotranslate/server/googleTranslate.ts [hunk 29]
```diff
@@ -85,37 +85,35 @@ class GoogleAutoTranslate extends AutoTranslate {
...
+			if (!request.ok && request.status === 400 && request.statusText === 'INVALID_ARGUMENT') {
+				throw new Error('Failed to fetch supported languages');
```
[the remark this answers is not in the document]
Comment (KevLehman) [reply apps/meteor/app/autotranslate/server/googleTranslate.ts:97]: This keeps the original logic

File: apps/meteor/app/federation/server/lib/http.js (hunk 16 above)
Comment (KevLehman) [reply apps/meteor/app/federation/server/lib/http.js:22]: Nah, not really necessary

File: apps/meteor/app/slackbridge/server/SlackAPI.js (hunk 25 above)
Comment (KevLehman) [reply apps/meteor/app/slackbridge/server/SlackAPI.js:147]: There's no actual perf reasons to not use backticks, but I'll change them :)

**Review by MarcosSpessatto — APPROVED** [review]
